statsmodels/survey/summary_stats.py

```diff
@@ -0,0 +1,538 @@
+"""
+Methods for creating summary statistics and their SE for survey data.
+
+The main classes are:
+
+  * SurveyDesign : Parent class that creates attributes for easy
+  implementation of other methods. Attributes include relabeled
+  clusters, number of clusters per strata, etc.
+
+  * SurveyStat : implements methods to calculate the standard
+  error of each statistic via either the bootstrap or jackknife
+
+  * SurveyMean : Calculates the mean of each column
+
+  * SurveyTotal : Calculates the total of each column
+
+  * SurveyQuantile: Calculates the specified quantile[s] of each column
+"""
+
+import numpy as np
+# import pandas as pd
+
+
+class SurveyDesign(object):
+    """
+    Description of a survey design, used by most methods
+    implemented in this module.
+
+    Parameters
+    -------
+    strata : array-like or None
+        Strata for each observation. If none, an array
+        of ones is constructed
+    cluster : array-like or None
+        Cluster for each observation. If none, an array
+        of ones is constructed
+    weights : array-like or None
+        The weight for each observation. If none, an array
+        of ones is constructed
+    nest : boolean
+        allows user to specify if PSU's with the same
+        PSU number in different strata are treated as distinct PSUs.
+
+    Attributes
+    ----------
+    weights : (n, ) array
+        The weight for each observation
+    nstrat : integer
+        The number of district strata
+    sclust : (n, ) array
+        The relabeled cluster array from 0, 1, ..
+    strat : (n, ) array
+        The related strata array from 0, 1, ...
+    ncs : (self.nstrat, ) array
+        Holds the number of clusters in each stratum
+    sfclust : ndarray
+        The stratum for each cluster
+    nclust : integer
+        The total number of clusters across strata
+    """
+
+    def __init__(self, strata=None, cluster=None, weights=None, rep_weights=None, nest=True):
```

Comment:
We might want to consider an alternative design, in which the `SurveyDesign` object constructs the `rep_weights` (either jack or boot), and the `SurveyStat` class obtains them from the `SurveyDesign` class. This has a few advantages:

* It separates the weight construction from the calculation of standard errors

* It unifies the setting where weights are known and where weights are calculated

* We could provide convenience methods for importing/exporting weights.

A common situation is that the data owner creates the rep_weights, then distributes the rep_weights to others without providing psu and strata.  This code should be usable by people who want to create rep_weights for others to use, and also by people who have rep_weights but do not have psu/strata.